library/include/BitArray.h

```diff
@@ -232,4 +234,299 @@ namespace DFHack
             m_data[idx] = item;
         }
     };
+
+    template <typename L, typename I>
```

Comment:
If you want this could just take a single template parameter and then require example requirements would be
<pre>
template&#60;typename ValueType&#62;
struct DfLinkedList {
using link_type = DfLinkedList&#60;ValueType&#62;::type;
};
// This would then be required only once for each linked list type
template<>
struct DfLinkedListLinkType&#60;job*&#62; {
using type = job_list_link;
};
// List could be then declared with familiar syntax.
DfLinkedList&#60;job*&#62; list;
</pre>

Reply:
Since the item field is always named `item`, we could also use `decltype(L::item)`.

Reply:
Actually, after trying it, we can't use `std::remove_pointer<decltype(L::item)>::type` because `L` is just the name of a class at the point when that code would be evaluated by the compiler and there aren't any fields associated with that class yet at that point.

Reply:
That sounds like an alternative. Only issue to me is that I feel link_type should be implementation detail that isn't "exposed" outside wrapper class.

Reply:
using value_type = std::remove_pointer<decltype(L{}.item)>::type; might be the syntax you need to make it work. But link_type should be always fully available type when template class is instantiated in the code using it. If it isn't fully defined type then whole class would fail. `DfLinkedListLinkType<job*> list;`  should be replacement for `job_list_link list;` where both require job_list_link to be fully defined.